README.md
django-taggit
=============

``django-taggit`` is a simpler approach to tagging with Django.

Getting Started
---------------

You can simply install it with
``pip``::

    $ pip install django-taggit


Add ``"taggit"`` to your
``INSTALLED_APPS`` then just add a TaggableManager to your model and go::

    from django.db import models

    from taggit.managers import TaggableManager

    class Food(models.Model):
        # ... fields here

        tags = TaggableManager()

Then you can use the API like so::

    >>> apple = Food.objects.create(name="apple")
    >>> apple.tags.add("red", "green", "delicious")
    >>> apple.tags.all()
    [<Tag: red>, <Tag: green>, <Tag: delicious>]
    
    >>> apple.tags.remove("green")
    >>> apple.tags.all()
    [<Tag: red>, <Tag: delicious>]
    
    >>> Food.objects.filter(tags__name__in=["red"])
    [<Food: apple>, <Food: cherry>]

Tags will show up for you automatically in forms and the admin.

If you want to enforce lowercase tags everywhere (recommended, to avoid
ending up with tags 'Music' and 'music' which are functionally identical
but show up in different taxonomies), add to settings.py:

``TAGGIT_FORCE_LOWERCASE = True``

If you want to prevent certain words from being added as tags (such as
English articles to, from, the, of, etc.) add to settings.py:

``TAGGIT_STOPWORDS = [u'a', u'an', u'and', u'be', u'from', u'of']``

Requirements
------------

``django-taggit`` requires Django 1.1 or greater.

Further Infos
-------------

For more info checkout out the
[documentation](http://readthedocs.org/docs/django-taggit/en/latest/).

And for questions about usage or development you can contact
the [mailinglist](http://groups.google.com/group/django-taggit).

Merger's notes
--------------

* Decided to stay with version 0.9.4 changed by shacker.
* Decided to leave markdown version of README.
* Decided not to merge forks:
    * extending base functionality or ones that would lower performance;
    * being much against PEP-8;
    * with changes solving specific problems.
* Combined alex/django-taggit & fcurella/django-taggit-autocomplete by AGoodId (https://github.com/AGoodId/django-taggit).


docs/changelog.txt
Changelog
=========

0.9.4
~~~~~

Unreleased.

 * *Backwards incompatible*  The ``name`` field of the ``Tag`` model is
   now marked as unique. You should update your database schema using
   ``ALTER TABLE taggit_tag ADD UNIQUE (name)``.

0.9.2
~~~~~

Unreleased.

 * *Backwards incompatible*  Forms containing a :class:`TaggableManager` by
   default now require tags, to change this provide ``blank=True`` to the
   :class:`TaggableManager`.

0.9.0
~~~~~

 * Added a Hebrew locale.
 * Added an index on the ``object_id`` field of ``TaggedItem``.
 * When displaying tags always join them with commas, never spaces.
 * The docs are now available `online <http://django-taggit.readthedocs.org/>`_.
 * Custom ``Tag`` models are now allowed.
 * *Backwards incompatible*  Filtering on tags is no longer
   ``filter(tags__in=["foo"])``, it is written
   ``filter(tags__name__in=["foo"])``.
 * Added a German locale.
 * Added a Dutch locale.
 * Removed ``taggit.contrib.suggest``, it now lives in an external application,
   see :doc:`external_apps` for more information.

0.8.0
~~~~~

 * Fixed querying for objects using ``exclude(tags__in=tags)``.
 * Marked strings as translatable.

   * Added a Russian translation.
 * Created a `mailing list <http://groups.google.com/group/django-taggit>`_.
 * Smarter tagstring parsing for form field; ported from Jonathan
   Buchanan's `django-tagging
   <http://django-tagging.googlecode.com>`_. Now supports tags
   containing commas. See :ref:`tags-in-forms` for details.
 * Switched to using savepoints around the slug generation for tags.  This
   ensures that it works fine on databases (such as Postgres) which dirty a
   transaction with an ``IntegrityError``.
 * Added Python 2.4 compatibility.
 * Added Django 1.1 compatibility.


taggit/managers.py
from django.contrib.contenttypes.generic import GenericRelation
from django.contrib.contenttypes.models import ContentType
from django.db import models
from django.db.models.fields.related import ManyToManyRel, RelatedField
from django.db.models.related import RelatedObject
from django.db.models.fields.related import add_lazy_relation
from django.utils.translation import ugettext_lazy as _
from django.utils.encoding import force_unicode

from taggit import settings
from taggit.forms import TagField
from taggit.models import TaggedItem, GenericTaggedItemBase
from taggit.utils import require_instance_manager
from taggit.widgets import TagAutocomplete


try:
    all
except NameError:
    # 2.4 compat
    try:
        from django.utils.itercompat import all
    except ImportError:
        # 1.1.X compat
        def all(iterable):
            for item in iterable:
                if not item:
                    return False
            return True


class TaggableRel(ManyToManyRel):
    def __init__(self, related_name=None):
        self.related_name = related_name
        self.limit_choices_to = {}
        self.symmetrical = True
        self.multiple = True
        self.through = None


class TaggableManager(RelatedField):
    def __init__(self, verbose_name=_("Tags"),
        help_text=_("A comma-separated list of tags."), through=None, blank=False, related_name=None):
        self.through = through or TaggedItem
        self.rel = TaggableRel(related_name)
        self.verbose_name = verbose_name
        self.help_text = help_text
        self.blank = blank
        self.editable = True
        self.unique = False
        self.creates_table = False
        self.db_column = None
        self.choices = None
        self.serialize = False
        self.null = True
        self.creation_counter = models.Field.creation_counter
        models.Field.creation_counter += 1

    def __get__(self, instance, model):
        if instance is not None and instance.pk is None:
            raise ValueError("%s objects need to have a primary key value "
                "before you can access their tags." % model.__name__)
        manager = _TaggableManager(
            through=self.through, model=model, instance=instance
        )
        return manager

    def contribute_to_class(self, cls, name):
        self.name = self.column = name
        self.model = cls
        cls._meta.add_field(self)
        setattr(cls, name, self)
        if not cls._meta.abstract:
            if isinstance(self.through, basestring):
                def resolve_related_class(field, model, cls):
                    self.through = model
                    self.post_through_setup(cls)
                add_lazy_relation(
                    cls, self, self.through, resolve_related_class
                )
            else:
                self.post_through_setup(cls)

    def post_through_setup(self, cls):
        self.use_gfk = (
            self.through is None or issubclass(self.through, GenericTaggedItemBase)
        )
        self.rel.to = self.through._meta.get_field("tag").rel.to
        if self.use_gfk:
            tagged_items = GenericRelation(self.through)
            tagged_items.contribute_to_class(cls, "tagged_items")

    def save_form_data(self, instance, value):
        getattr(instance, self.name).set(*value)

    def formfield(self, form_class=TagField, **kwargs):
        defaults = {
            "label": self.verbose_name,
            "help_text": self.help_text,
            "required": not self.blank,
        }
        if settings.TAGGIT_AUTOCOMPLETE_WIDGET:
            defaults.extend({
                "widget": TagAutocomplete,
            })
        defaults.update(kwargs)
        
        return form_class(**defaults)

    def value_from_object(self, instance):
        if instance.pk:
            return self.through.objects.filter(**self.through.lookup_kwargs(instance))
        return self.through.objects.none()

    def related_query_name(self):
        return self.model._meta.module_name

    def m2m_reverse_name(self):
        return self.through._meta.get_field_by_name("tag")[0].column
    
    def m2m_reverse_field_name(self):
        return self.through._meta.get_field_by_name("tag")[0].name

    def m2m_target_field_name(self):
        return self.model._meta.pk.name

    def m2m_reverse_target_field_name(self):
        return self.rel.to._meta.pk.name

    def m2m_column_name(self):
        if self.use_gfk:
            return self.through._meta.virtual_fields[0].fk_field
        return self.through._meta.get_field('content_object').column

    def db_type(self, connection=None):
        return None

    def m2m_db_table(self):
        return self.through._meta.db_table

    def extra_filters(self, pieces, pos, negate):
        if negate or not self.use_gfk:
            return []
        prefix = "__".join(["tagged_items"] + pieces[:pos-2])
        cts = map(ContentType.objects.get_for_model, _get_subclasses(self.model))
        if len(cts) == 1:
            return [("%s__content_type" % prefix, cts[0])]
        return [("%s__content_type__in" % prefix, cts)]

    def bulk_related_objects(self, new_objs, using):
        return []


class _TaggableManager(models.Manager):
    def __init__(self, through, model, instance):
        self.through = through
        self.model = model
        self.instance = instance
        self.force_lowercase = getattr(settings, 'TAGGIT_FORCE_LOWERCASE', False)

    def get_query_set(self):
        return self.through.tags_for(self.model, self.instance)

    def _lookup_kwargs(self):
        return self.through.lookup_kwargs(self.instance)

    @require_instance_manager
    def add(self, *tags):
        if self.force_lowercase:
            lower_tags = []
            for t in tags:
                if not isinstance(t, self.through.tag_model()):
                    t = t.lower()
                lower_tags.append(t)
            tags = lower_tags

        tags = [force_unicode(t) if not isinstance(t, self.through.tag_model()) else t for t in tags]
        str_tags = set([
            t
            for t in tags
            if not isinstance(t, self.through.tag_model())
        ])

        tag_objs = set(tags) - str_tags
        if len(str_tags) == 0:
            existing = self.through.tag_model().objects.none()
        else: # above check for length of str_tags avoids query here
            existing = self.through.tag_model().objects.filter(
                name__regex=r'(^' + '$|^'.join(str_tags) + '$)'
            )
        tag_objs.update(existing)
        
        existing_names = set(t.name for t in existing)
        existing_names_lower = set(t.name.lower() for t in existing)
        
        for new_name in str_tags - existing_names:
            if len(set([new_name.lower()]) - existing_names_lower) > 0:
                tag_objs.add(self.through.tag_model().objects.create(name=new_name))

        for tag in tag_objs:
            self.through.objects.get_or_create(tag=tag, **self._lookup_kwargs())

    @require_instance_manager
    def set(self, *tags):
        have = set(tag.name for tag in self.get_query_set().all())
        wanted = set([tag.name if isinstance(tag, self.through.tag_model()) else tag for tag in tags])
        
        add = wanted - have
        remove = have - wanted
        
        self.add(*list(add))
        self.remove(*list(remove))

    @require_instance_manager
    def remove(self, *tags):
        if tags:
            self.through.objects.filter(**self._lookup_kwargs()).filter(
                tag__name__regex=r'(^' + '$|^'.join(tags) + '$)').delete()

    @require_instance_manager
    def clear(self):
        self.through.objects.filter(**self._lookup_kwargs()).delete()

    def most_common(self):
        return self.get_query_set().annotate(
            num_times=models.Count(self.through.tag_relname())
        ).order_by('-num_times')

    @require_instance_manager
    def similar_objects(self, num=None, **filters):
        lookup_kwargs = self._lookup_kwargs()
        lookup_keys = sorted(lookup_kwargs)
        qs = self.through.objects.values(*lookup_kwargs.keys())
        qs = qs.annotate(n=models.Count('pk'))
        qs = qs.exclude(**lookup_kwargs)
        subq = self.all()
        qs = qs.filter(tag__in=list(subq))
        qs = qs.order_by('-n')
        
        if filters is not None:
            qs = qs.filter(**filters)
    
        if num is not None:
            qs = qs[:num]
        
        # TODO: This all feels like a bit of a hack.
        items = {}
        if len(lookup_keys) == 1:
            # Can we do this without a second query by using a select_related()
            # somehow?
            f = self.through._meta.get_field_by_name(lookup_keys[0])[0]
            objs = f.rel.to._default_manager.filter(**{
                "%s__in" % f.rel.field_name: [r["content_object"] for r in qs]
            })
            for obj in objs:
                items[(getattr(obj, f.rel.field_name),)] = obj
        else:
            preload = {}
            for result in qs:
                preload.setdefault(result['content_type'], set())
                preload[result["content_type"]].add(result["object_id"])

            for ct, obj_ids in preload.iteritems():
                ct = ContentType.objects.get_for_id(ct)
                for obj in ct.model_class()._default_manager.filter(pk__in=obj_ids):
                    items[(ct.pk, obj.pk)] = obj

        results = []
        for result in qs:
            obj = items[
                tuple(result[k] for k in lookup_keys)
            ]
            obj.similar_tags = result["n"]
            results.append(obj)
        return results


def _get_subclasses(model):
    subclasses = [model]
    for f in model._meta.get_all_field_names():
        field = model._meta.get_field_by_name(f)[0]
        if (isinstance(field, RelatedObject) and
            getattr(field.field.rel, "parent_link", None)):
            subclasses.extend(_get_subclasses(field.model))
    return subclasses
